File: solutions/c/resistor-color-trio/1/resistor_color_trio.c
#include "resistor_color_trio.h"

resistor_value_t color_code(resistor_band_t* bands){
  resistor_value_t res = {
    .value = 0,
    .unit = OHMS
  };
  long value = bands[0] * 10 + bands[1];
  long unit = 1;
  for(int i = 0; i < (int)bands[2]; i++) unit *= 10;
  value *= unit;
  if(value < 1000) {
    res.unit = OHMS;
    res.value = value;
  }
  else if(value < 1000000) {
    res.unit = KILOOHMS;
    res.value = value / 1000;
  }
  else if(value < 1000000000){
    res.unit = MEGAOHMS;
    res.value = value / 1000000;
  }
  else if(value < 1000000000000){
    res.unit = GIGAOHMS;
    res.value = value / 1000000000;
  }
  return res;
}

File: solutions/csharp/resistor-color-trio/1/ResistorColorTrio.cs
public static class ResistorColorTrio
{
    enum ResistorColor {
        Black,
        Brown,
        Red,
        Orange,
        Yellow,
        Green,
        Blue,
        Violet,
        Grey,
        White
    }
    
    public static string Label(string[] colors)
    {
        long first = (long)ParseResistorColor(colors[0]);
        long second = (long)ParseResistorColor(colors[1]);
        long multiplier = (long)ParseResistorColor(colors[2]);
    
        long res = (first * 10 + second) * (long)Math.Pow(10, multiplier);
    
        if (res >= 1_000_000_000) return $"{res / 1_000_000_000} gigaohms";
        if (res >= 1_000_000) return $"{res / 1_000_000} megaohms";
        if (res >= 1_000) return $"{res / 1_000} kiloohms";
        return $"{res} ohms";
    }


    private static ResistorColor ParseResistorColor(string color)
    {
        if (!Enum.TryParse(color, true, out ResistorColor result))
            throw new ArgumentException($"Invalid resistor color: {color}");
        return result;
    }
}
